File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeVectorFilter.kt
package io.github.rajibsarwar.edgeframeworks

data class EdgeVectorFilter(
    val documentId: String? = null,
    val collectionId: String? = null,
    val metadata: Map<String, String> = emptyMap()
) {
    fun matches(chunk: EdgeChunk): Boolean {
        if (
            documentId != null &&
            chunk.documentId != documentId
        ) {
            return false
        }

        if (
            collectionId != null &&
            chunk.metadata["collectionID"] != collectionId
        ) {
            return false
        }

        return metadata.all { (key, value) ->
            chunk.metadata[key] == value
        }
    }
}

File: ios/Sources/EdgeFrameworks/Core/EdgeVectorFilter.swift
public struct EdgeVectorFilter: Equatable, Sendable {
    public let documentID: String?
    public let collectionID: String?
    public let metadata: [String: String]

    public init(
        documentID: String? = nil,
        collectionID: String? = nil,
        metadata: [String: String] = [:]
    ) {
        self.documentID = documentID
        self.collectionID = collectionID
        self.metadata = metadata
    }

    public func matches(_ chunk: EdgeChunk) -> Bool {
        if let documentID, chunk.documentID != documentID {
            return false
        }

        if let collectionID,
           chunk.metadata["collectionID"] != collectionID {
            return false
        }

        for (key, value) in metadata
        where chunk.metadata[key] != value {
            return false
        }

        return true
    }
}
